packages/prisma/src/scripts/2024-08-23_compute_instance_statistics.ts

```diff
@@ -0,0 +1,234 @@
+        const { tempCorrect1, tempIncorrect1 } = Object.entries(
+          instance.anonymousResults.choices
+        ).reduce(
+          (acc, [ix, count]) => {
+            if (ix === correctAnswerIx) {
+              acc.tempCorrect1 += count as number
+            } else {
+              acc.tempIncorrect1 += count as number
+            }
+
+            return acc
+          },
+          { tempCorrect1: 0, tempIncorrect1: 0 }
+        )
+
+        anonymousCorrectCount = tempCorrect1
+        anonymousWrongCount = tempIncorrect1
+        break
+
+      case ElementType.NUMERICAL:
+      case ElementType.FREE_TEXT:
+        const { tempCorrect2, tempIncorrect2 } =
+          instance.anonymousResults.responses.reduce(
+            (acc, response) => {
+              acc.tempCorrect += Number(response.correct)
+              acc.tempIncorrect += Number(!response.correct)
+              return acc
+            },
+            { tempCorrect: 0, tempIncorrect: 0 }
+          )
+
+        anonymousCorrectCount = tempCorrect2
+        anonymousWrongCount = tempIncorrect2
+        break
+
+      case ElementType.MC:
+      case ElementType.KPRIM:
+        console.log(
+          'Anonymous results for MC/KPRIM cannot be recovered, the corresponding counts are set to -1'
+        )
+        anonymousCorrectCount = -1
+        anonymousPartialCorrectCount = -1
+        anonymousWrongCount = -1
+
+      default:
+        throw new Error(`Unknown element type ${instance.elementData.type}`)
+    }
+
+    // update the instance statistics
+    return prisma.elementInstance.update({
+      where: { id: instance.id },
+      data: {
+        instanceStatistics: {
+          update: {
+            // set correctness counts for all statistics with responses
+            correctCount,
+            partialCorrectCount: partialCount,
+            wrongCount,
+
+            // set first and last correctness counts for practice quiz
+            firstCorrectCount: isPracticeQuiz ? firstCorrectCount : undefined,
+            firstPartialCorrectCount: isPracticeQuiz
+              ? firstPartialCount
+              : undefined,
+            firstWrongCount: isPracticeQuiz ? firstWrongCount : undefined,
+            lastCorrectCount: isPracticeQuiz ? lastCorrectCount : undefined,
+            lastPartialCorrectCount: isPracticeQuiz
+              ? lastPartialCount
+              : undefined,
+            lastWrongCount: isPracticeQuiz ? lastWrongCount : undefined,
+
+            // update anonymous correctness counts
+            anonymousCorrectCount,
+            anonymousPartialCorrectCount,
+            anonymousWrongCount,
+
+            // set average time spent on instance and total unique participants
+            uniqueParticipantCount: totalUniqueParticipants,
+            averageTimeSpent: averageInstanceTime,
+          },
+        },
+      },
+    })
+  })
+
+  await prisma.$transaction(updates)
```

Comment:
are all updates to all instances in one transaction like this? I would rather not do such a large transaction and process them separately

Reply:
yes, this was not ideal - leftover from copied script :) fixed it